### ISSUE TITLE
Resource inputs not populated during reads for import operations

### ISSUE THREAD

**Issue by praneetloke** [pulumi/pulumi#11279]
### What happened?

(I am not sure if this should be a bug report or a feature request.)

I am working on a [Render provider](https://github.com/cloudy-sky-software/pulumi-render) that uses REST APIs with path params. I've modeled the path params as required inputs for the resources. This works just fine until I have to import a resource that has a path param which represents the ID of a parent resource. 

For example, take the Custom Domain resource which is always associated with a Service resource, so the endpoint for custom domains has the service ID as one of its path params. When [importing a custom domain resource](https://api-docs.render.com/reference/get-custom-domain), the `ReadRequest` param that the `Read` method is called with does not have any `inputs` (the current inputs for the resource) or `properties` (aka the existing state, which is understandable because it does not exist in the state.) That makes it impossible to satisfy the required path param as there is no other way to pass the user supplied resource inputs to the `Read` method.

The [doc comment](https://github.com/pulumi/pulumi/blob/master/sdk/proto/go/provider.pb.go#L1332) for the `Inputs` property in the `ReadRequest` indicates that it is only populated for refresh operations. Is there a reason that that is the case? If the `Inputs` property in the `ReadRequest` were populated with the current inputs of the resource, that would solve this issue.

Here's the resource I am trying to import:

```
const domain = new render.services.CustomDomain(
    "wwwCustomDomain",
    {
        name: "www.example.com",
        serviceId: staticSite.id, // If Inputs were populated, I would have been able to use this value to fill-in the required path param of the parent service ID in the endpoint to read a custom domain resource.
    },
    { import: "some-id" } // The import ID is the other path param which is the unique provider-assigned ID of the resource itself.
);
```

Here's how I am unmarshaling the inputs from the `ReadRequest`, just in case there is a problem with how I am doing that:

```go
inputs, err := plugin.UnmarshalProperties(req.GetInputs(), plugin.MarshalOptions{KeepUnknowns: true, KeepSecrets: true, SkipNulls: true})
// inputs is empty here.
```

Hope I am not missing something here. Any help is greatly appreciated.

### Steps to reproduce

N/A

### Expected Behavior

`Inputs` in `ReadRequest` is populated with the resource's inputs for the purposes of importing the resource.

### Actual Behavior

`Inputs` is empty in the `ReadRequest` for a resource that is being imported.

### Output of `pulumi about`

I am using the latest version of `pulumi/pulumi/pkg` and `pulumi/pulumi/sdk` which, at the time of writing, is `3.46.0`.

### Contributing

Vote on this issue by adding a 👍 reaction. 
To contribute a fix for this issue, leave a comment (and link to your pull request, if you've opened one already).

**Comment by kpitzen** [comment]
Hi @praneetloke !  I appreciate the amount of thought that went into this issue - the detail is much appreciated.  It seems like the current functionality is as-intended, but we should certainly have some discussions around where there may be room to help support this specific use case.  I'll discuss with a couple of people and see if we can come up with a prompt answer for you.  Thanks again!

**Comment by kpitzen** [comment]
@praneetloke After chatting with the team, we don't see any immediate reason to not remove the restriction for inputs to be available only within refresh.  The initial optionality was likely there to not place an undue burden around providing this information in the future, but your use-case seems more than compelling enough to provide it :)

**Comment by Frassle** [comment]
Echo'ing Kyle above it looks like it might be fine to start passing inputs during imports not just refresh, although we need to give a bit of a think of if we also need to support that for the `pulumi import` command.

Possibly a better solution (and the standard way of solving this) is to put everything needed for an import into the ID of the resource. So rather than the ID just being the one path param you would instead combine both path params into one string with some safe delimiter between them, something like "static_site_id/some_id".

**Comment by praneetloke** [comment]
> Possibly a better solution (and the standard way of solving this) is to put everything needed for an import into the ID of the resource. So rather than the ID just being the one path param you would instead combine both path params into one string with some safe delimiter between them, something like "static_site_id/some_id".

I actually thought of doing this as a workaround but felt it would be a bit of a hack that _may_ not work well in all conditions. It would require provider implementations to get creative with handling the formatted ID in ways that might be brittle. My thinking around asking for the inputs to be populated is that it would allow implementations to programmatically pluck whichever inputs they need based on the API definition.

**Comment by praneetloke** [comment]
Ping. Would love to see this in a nearby release. 😁

**Comment by vmpn** [comment]
Ran into same issue with importing cloudflare email routing rules as it needs zone_id parameter besides id.

What needs to be changed to support this?

**Comment by praneetloke** [comment]
Hello 👋  I was just wondering if there are concerns about implementing this change.

**Comment by praneetloke** [comment]
@Frassle wanted to see if this was something you would accept a PR for? If so, I am happy to looking into submitting one myself.

**Comment by Frassle** [comment]
> wanted to see if this was something you would accept a PR for?

Yup there's just two design worries I have around it. Which if we had answer for would make the PR a lot easier to write and approve.

Firstly what do we do about the `pulumi import` command. I am not going to accept something where we design in that some resource don't work with that command, so we need something to allow inputs to that. When using the JSON file input for it there's a fairly obvious solution of just putting the properties as an extra field in the JSON. So its mostly what's a sensible design for if people are just importing one thing with like a property to set? This is mostly a UX problem more than anything else, but unusable because the UX is so bad isn't much better than unusable because it's just not implemented.

Secondly, we need some assurance from Providers (cc @kpitzen & @AaronFriel) around how providers will behave if they start getting properties&inputs filled in during Refresh. I suspect it _should_ be fine, but this is one of the less well know parts of the system.

**Comment by praneetloke** [comment]
@Frassle I noticed that #5973 was recently closed and also came across #14115. I wanted to bump this use case up to include inputs for the `Read` provider method as well.

Since you've added a flag in `Configure` to allow the provider to opt-in to the new behavior, I feel like it should be ok to include the `inputs` for `Read` as well. Curious what you think.

**Comment by praneetloke** [comment]
Giving this issue a 2024 bump since #14115 has been merged as well.

**Comment by Frassle** [comment]
So I think the remaining question here isn't one of grpc interface but of user interface.

The gRCP request _already_ contains "inputs", the problem is that currently the only time the engine has inputs to send is for a refresh operation (where we use the inputs currently saved in state).

For an import there are two case:
1. Using the import option. In which case we _could_ send the input properties from the program.
2. Using the import command. Which we'd need a new thing on the CLI for the user to send the inputs to the engine for it to send on to the provider.

Scenario 2 is the tricky one. I don't want to just support option 1 because that could lead to resources that you can only import via the resource option, and are not importable via the command line.

**Comment by praneetloke** [comment]
For scenario 2, wouldn't the inputs be `nil` since the only thing you can supply if the ID of the resource being imported via the CLI? Or am I misunderstanding something about that import method?

**Comment by Frassle** [comment]
Indeed, that's my point. If a you write a resource import that depends on inputs then users wouldn't be able to import it with the import command any more.

**Comment by praneetloke** [comment]
I see what you are saying. But do you agree that it is understood that certain resources cannot be imported via just the command-line since they need more than just the ID in order to be imported successfully?

**Comment by Frassle** [comment]
I think we should be very careful about making a decision that locks the system into that behaviour. 

Resources are supposed to be importable by their import string, which is normally just their ID but the import string doesn't have to be just the ID. Today resources that need more just find a way to encode that extra property data into the import string with something like "id/extra". 

This works today, it works for the import option and the command and the only downside is as you described above "It would require provider implementations to get creative with handling the formatted ID in ways that might be brittle.".

So I'm very happy to consider what we could add to make this easier for providers to implement but it doesn't seem worth it to me if it's at the cost of some resources now only being importable via the resource option and not the command.

**Comment by praneetloke** [comment]
@Frassle Just wanted to post a quick update that I did go the route of using segmented IDs to support importing resources that are backed by a RESTful API as part of [`pulumi-provider-framework`](https://github.com/cloudy-sky-software/pulumi-provider-framework/). I think for now my use case is satisfied. It's possible that in the future there might be a different reason to deliver the inputs during refresh, so I'll leave this issue open. But if you feel you would like to close it, please feel free to do so.
